Disable support for snapshot id within the Iceberg table name

findinpath opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->
Disable the possibility of specifying the snapshot id
within the Iceberg table name.

With the introduction of the syntax

```
table1 FOR VERSION AS OF 123
```
there is no more need of supporting the
deprecated syntax:

```
table1@123
```

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?

Other.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Iceberg connector

> How would you describe this change to a non-technical end user or system administrator?

Use `table1 FOR VERSION AS OF 123` instead of `table1@123` for retrieving the snapshot `123` of `table`.

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Disable support for specifying the snapshot id within the Iceberg table name
```

findinpath commented:
@findepi @phd3 @electrum PTAL on this PR

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -312,13 +307,7 @@ public IcebergTableHandle getTableHandle(
-        }
-
-        Optional<Long> snapshotId = endVersion.map(version -> getSnapshotIdFromVersion(table, version))
-                .or(() -> getSnapshotId(table, name.getSnapshotId(), isAllowLegacySnapshotSyntax(session)));
```
This used to be doing `Optional.ofNullable(table.currentSnapshot()).map(Snapshot::snapshotId)`.
Was the removal intentional?

alexjo2144 replied:
We need that

findinpath replied:
Appologies. The removal was not intentional.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:
```diff
@@ -292,20 +293,6 @@ public double getMinimumAssignedSplitWeight()
         return minimumAssignedSplitWeight;
     }
 
-    @Config("iceberg.allow-legacy-snapshot-syntax")
```
Add a `DefunctConfig` annotation

findinpath replied:
Already done https://github.com/trinodb/trino/pull/13414/files/80a5cca73dda556d315878952a11fe1fae8e62dc#diff-80ab7495fcc64c0a92d83a014d8c56697fb3fe09f67a9edd9e3da116264b7502R39

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -403,21 +392,15 @@ private Optional<SystemTable> getRawSystemTable(ConnectorSession session, Schema
                 return Optional.of(new SnapshotsTable(systemTableName, typeManager, table));
             case PARTITIONS:
-                return Optional.of(new PartitionTable(systemTableName, typeManager, table, getSnapshotId(table, name.getSnapshotId(), isAllowLegacySnapshotSyntax(session))));
+                return Optional.of(new PartitionTable(systemTableName, typeManager, table, Optional.empty()));
```
This looks like a regression to me, does the `AS OF` syntax work as a replacement for `select * from abc$manifests@456`?

findinpath replied:
I have a PR for supporting time travel queries on system tables but it is unfortunately not ready yet.

findinpath replied:
Probably it is best to introduce the ability to perform time travel queries on system tables before continuing with this PR.

findinpath replied:
As shown in the PRs: 

- https://github.com/trinodb/trino/pull/13497
- https://github.com/trinodb/trino/pull/13854

time travel on system tables incurs added complexity to the Iceberg connector code.

Without concrete demand from the community for time travel functionality on system tables, we leave the above mentioned PRs on hold and go further with disabling support  for snapshot id within the Iceberg table name.

findepi replied:
> This looks like a regression to me

correct

> does the `AS OF` syntax work as a replacement for `select * from abc$manifests@456`?

no

The reason we're _not_ pursing the above linked PRs is because we believe there is little community interest in these.
If there isn't general demand on time travel on system tables, we should be OK with carrying out removal of this (already deprecated) functionality.

findinpath commented:
In the context of deprecating the `tableName@snapshotId` syntax, maybe it is worth also rewriting `io.trino.plugin.iceberg.IcebergTableHandle#toString` to use `tableName FOR VERSION AS OF snapshotId` syntax.

findepi commented:
@findinpath there is a test failure. is it related?

findepi commented:
> In the context of deprecating the `tableName@snapshotId` syntax, maybe it is worth also rewriting `io.trino.plugin.iceberg.IcebergTableHandle#toString` to use `tableName FOR VERSION AS OF snapshotId` syntax.

This is used in EXPLAIN.
Let's keep it terse (no change), since this will always be there, for every SELECT, not only for time travel queries.

findinpath commented:
@findepi  failure of the tests was related to my changes. I fixed the failing test failure and rebased on `master` due to conflicts which appeared in the meantime.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -332,16 +327,11 @@ public IcebergTableHandle getTableHandle(
-            long snapshotId = endVersion.map(connectorTableVersion -> getSnapshotIdFromVersion(table, connectorTableVersion))
-                    .orElseGet(() -> resolveSnapshotId(table, name.getSnapshotId().get(), isAllowLegacySnapshotSyntax(session)));
+        if (endVersion.isPresent()) {
+            long snapshotId = endVersion.map(connectorTableVersion -> getSnapshotIdFromVersion(table, connectorTableVersion)).get();
```
```suggestion
            long snapshotId = getSnapshotIdFromVersion(table, endVersion.get());
```

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -2090,19 +2074,9 @@ public void setTableAuthorization(ConnectorSession session, SchemaTableName tabl
-    }
-
-    private long resolveSnapshotId(Table table, long id, boolean allowLegacySnapshotSyntax)
+    private Optional<Long> getSnapshotId(Table table)
```
`getCurrentSnapshotId`

findepi commented:
No release notes since the functionality was already deprecated.
